### PR TITLE
Implement sponge construction.

### PR DESCRIPTION
**Pull request by porcuquine**
This PR adds support for a generalized sponge construction. The sponge API is defined by `SpongeTrait` and implemented by both `Sponge` and `SpongeCircuit`. Sponges can be instantiated in either `Simplex` or `Duplex` modes.

This generalizes work begun in #127, and once this PR merges, I will update #127 to use these APIs. I have begun updating that PR to use these changes and have verified that the tests which record specific encryption results (as concrete field elements) still pass when doing so.

More tests are still needed, and the initial draft may still have some bugs or unhandled edge cases.

### CODE REVIEW

File: src/sponge.rs [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+    fn squeeze_elements(&mut self, count: usize, _acc: &mut ()) -> Vec<Self::Elt> {
+        self.take(count).collect()
+    }
+}
+
+impl<F: PrimeField, A: Arity<F>> Iterator for Sponge<'_, F, A> {
+    type Item = F;
+
+    fn next(&mut self) -> Option<F> {
+        self.squeeze(&mut ())
```
Comment (dignifiedquire) [inline src/sponge.rs:396]: uhmmmm, what?

Comment (porcuquine) [reply src/sponge.rs:396]: > uhmmmm, what?

When making the circuit, this will be a constraint-system namespace. Here, we don't need any bookkeeping other than to plumb the argument.

**Comment by porcuquine** [comment]
I marked this as ready for review. I would like to add further tests, but I think what is here now is sufficient for review.

File: src/poseidon.rs [hunk 2]
```diff
@@ -296,7 +295,7 @@ where
                 })
             }
         };
-        let width = preimage.len();
+        let width = preimage.len() + 1;
```
Comment (dignifiedquire) [inline src/poseidon.rs:298]: this looks supsicious, why this change?

Comment (porcuquine) [reply src/poseidon.rs:298]: It was wrong but never mattered before.

**Comment by emmorais** [comment]
Consider adding a link to some document containing the fundamentals, such as [this one](https://keccak.team/files/CSF-0.1.pdf).

File: src/sponge.rs [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+use std::collections::VecDeque;
+
+/*
+
+A sponge can be instantiated in either simplex or duplex mode. Once instantiated, a sponge's mode never changes.
+
+At any time, a sponge is operating in one of two directions: squeezing or absorbing. All sponges are initialized in the
+absorbing direction. The number of absorbed field elements is incremented each time an element is absorbed and
+decremented each time an element is squeezed. The count of currently absorbed elements can never decrease below zero, so
+only as many elements as have been absorbed can be squeezed at any time.
```
[the remark this answers is not in the document]
Comment (porcuquine) [reply src/sponge.rs:14]: You're right. That should only apply to the duplex mode. I'll adjust, thanks.

[the remark this answers is not in the document]
Comment (porcuquine) [reply src/sponge.rs]: @Kubuxu Please see changes here: [Don't limit squeezing in simplex mode.](https://github.com/filecoin-project/neptune/pull/151/commits/1b925ac2b91747a17d6a5d541934db75ddbf552c).

File: src/lib.rs [hunk 4]
```diff
@@ -53,6 +53,11 @@ mod preprocessing;
 mod round_constants;
 mod round_numbers;
 
+/// Sponge
+pub mod sponge;
+pub mod sponge_api;
+pub mod sponge_circuit;
```
Comment (DrPeterVanNostrand) [inline src/lib.rs:59]: Can these be moved into a single `sponge` module?

File: src/hash_type.rs [hunk 5]
```diff
@@ -21,32 +21,21 @@ pub enum HashType<F: PrimeField, A: Arity<F>> {
     ConstantLength(usize),
     Encryption,
     Custom(CType<F, A>),
+    Sponge,
 }
 
 impl<F: PrimeField, A: Arity<F>> HashType<F, A> {
-    pub fn domain_tag(&self, strength: &Strength) -> F {
```
Comment (DrPeterVanNostrand) [inline src/hash_type.rs:27]: Why does adding a sponge construction remove the strength component from the domain tag?

File: src/poseidon.rs [hunk 6]
```diff
@@ -343,35 +344,47 @@ where
 
         Ok(self.pos - 1)
     }
-
```
Comment (DrPeterVanNostrand) [inline src/poseidon.rs:346]: Add line break.

File: src/poseidon.rs [hunk 7]
```diff
@@ -741,55 +765,170 @@ mod tests {
             {
                 match test_arity {
                     2 => scalar_from_u64s([
-                        0x3abccd9afc5729b1,
-                        0x31662bb49883a7dc,
-                        0x2a0ae894f8500373,
-                        0x5f3027eb2ef4f4b8,
+                        0x793dbaf54552cd69,
```
Comment (DrPeterVanNostrand) [inline src/poseidon.rs:768]: These test vectors have changed because strength was removed from the domain tag?

File: src/lib.rs (hunk 4 above)
Comment (porcuquine) [reply src/lib.rs]: Yes, good point. Done.

File: src/hash_type.rs (hunk 5 above)
Comment (porcuquine) [reply src/hash_type.rs:27]: Ah, I forgot to update the description. I removed the strength component from the domain tag because it was adding unnecessary complexity. It's mostly orthogonal to addition of the sponge — apart from that I had to consider it as part of this work.

I think from a library perspective this is fine at least for known consumers (i.e. I am not aware of anyone actually relying on the concrete values of `Strength::Strengthened` hashes).

It's also fine because we don't need to use the DST to distinguish hashes of different strengths. Because they will have different round constants and round numbers, the permutations cannot collide anyway.

Obviously this needs to be a breaking change, but the upcoming release will be.

File: src/poseidon.rs (hunk 7 above)
Comment (porcuquine) [reply src/poseidon.rs:769]: Yes, as above.

File: src/sponge.rs [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+        let mut sponge = Sponge::new_with_constants(&c, Mode::Simplex);
+
+        let mut elements: Vec<F> = Vec::with_capacity(n);
+        for _ in 0..n {
+            elements.push(F::random(&mut *rng));
+        }
+
+        let acc = &mut ();
+
+        // Reminder: a duplex sponge should encode its length as a prefix.
```
[the remark this answers is not in the document]
Comment (porcuquine) [reply src/sponge.rs]: Yes, copy/paste error when writing the test. Will fix.

[the remark this answers is not in the document]
Comment (porcuquine) [reply src/sponge.rs]: Actually, this was indeed just a reminder that although not present in this simplex test, a duplex sponge should encode its length as a prefix (if such a length exists — or as repeated prefixes, one header per message, if appropriate).

File: src/hash_type.rs (hunk 5 above)
Comment (DrPeterVanNostrand) [reply src/hash_type.rs:27]: Ok, thanks for the explanation.

**Comment by porcuquine** [comment]
The authors of the Sponge API specification (https://hackmd.io/bHgsH6mMStCVibM_wYvb2w) have privately reviewed this PR, and it now seems to comply with their intentions — so I am going to go ahead and merge.

The next step in this line of work will be to update #127 to use the applicable constructs introduced here.
